src/transformers/modeling_flax_bert.py

```diff
@@ -0,0 +1,332 @@
+
+    def apply(self, hidden_state, attention_mask, num_heads: int, head_size: int, intermediate_size: int):
+        attention = BertAttention(hidden_state, attention_mask, num_heads, head_size, intermediate_size, name="attention")
+        intermediate = BertIntermediate(attention, intermediate_size, name="intermediate")
+        output = BertOutput(intermediate, attention, name="output")
+
+        return output
+
+
+class BertLayerCollection(nn.Module):
+    """
+    Stores N BertLayer(s)
+    """
+    def apply(self, input, attention_mask, num_layers: int, num_heads: int, head_size: int, intermediate_size: int):
+        assert num_layers > 0, "num_layers should be >= 1, got ({})".format(num_layers)
+
+        # Initialize input / output
+        input_i = output_i = input
+
+        # Forward over all encoders
+        for i in range(num_layers):
+            output_i = BertLayer(input_i, attention_mask, num_heads, head_size, intermediate_size, name="{}".format(i))
+            input_i = output_i
+        return output_i
+
+
+class BertEncoder(nn.Module):
+
+    def apply(self, hidden_state, attention_mask, num_layers: int, num_heads: int, head_size: int, intermediate_size: int):
+        encodings = BertLayerCollection(hidden_state, attention_mask, num_layers, num_heads, head_size, intermediate_size, name="layer")
+        return encodings
+
+
+class BertPooler(nn.Module):
+
+    def apply(self, hidden_state):
+        first_token = hidden_state[:, 0]
+        out = nn.Dense(first_token, hidden_state.shape[-1], name="dense")
+        return jax.lax.tanh(out)
+
+
+class BertModel(nn.Module):
+
+    def apply(self, input_ids, token_type_ids, position_ids, attention_mask,
+              vocab_size: int, hidden_size: int, type_vocab_size: int,
+              max_length: int, num_encoder_layers: int, num_heads: int,
+              head_size: int, intermediate_size: int, padding_idx: int,
+              emb_init: Callable[..., np.ndarray] = nn.initializers.normal(stddev=0.1)):
+
+        # Embedding
+        embeddings = BertEmbeddings(
+            input_ids, token_type_ids, position_ids, attention_mask,
+            vocab_size, hidden_size, type_vocab_size, max_length,
+            name="embeddings"
+        )
+
+        # N stacked encoding layers
+        encoder = BertEncoder(
+            embeddings, attention_mask, num_encoder_layers,
+            num_heads, head_size, intermediate_size,
+            name="encoder"
+        )
+
+        pooled = BertPooler(encoder, name="pooler")
+        return encoder, pooled
+
+
+class FlaxBertModel(JaxPreTrainedModel):
+    """
+    BERT implementation using JAX/Flax as backend
+    """
+
+    model_class = BertModel
+    config_class = BertConfig
+    pretrained_model_archive_map = BERT_PRETRAINED_MODEL_ARCHIVE_MAP
+    base_model_prefix = "bert"
+
+    @staticmethod
+    def convert_from_pytorch(pt_state: Dict, config: BertConfig) -> Dict:
+        jax_state = dict(pt_state)
+
```

Comment:
In the end, you've had to add a conversion script from PyTorch to Flax! Is this so you have more flax-idiomatic code?
We usually have methods for this at the beginning of each file, instead of static methods inside the models. What is the advantage of having those in the models?

As an aside, these transformations seem to be the exact opposite of those needed to go from TensorFlow to PyTorch. From this it seems that converting between Flax and TensorFlow will require less glue than from TF to PT or from PT to Flax!

Reply:
As a second point, will this method become more abstract down the road? It seems a bit painful to write for every conversion from PyTorch to Flax. 

You're probably aware of that but usually, it's best to think deeply about the model architecture (and therefore conversion script/method) before merging and uploading weights on the S3, as we try to never modify the weights on the S3. Having a good architecture at first is important, and so is making sure that this conversion method isn't too much work for each subsequent architecture.